### PR TITLE
[MERGE] project: link milestones in tasks

### PR DESCRIPTION
**Pull request by xavierbol**
Purpose
=======

The goal of this PR is to create a link between milestones and tasks. Indeed, milestones usually contain a set of tasks. Once all of these tasks are done, the milestone can be considered as reached. This link would help users organize their projects more efficiently and have a better overview of what needs to be done when.

## Implementation details

- A new many2one field is added in the `project.task` called `milestone_id` to link a milestone to tasks.
- Add `Milestones` as global setting and project setting. It means the user will can globally enable the feature and choose in which project he wants to use the feature.
- In the form view of milestone, a stat button is added to display the tasks linked to current milestone.
- Add `Late Milestone` filter in tasks, when this filter is applied the user will see the tasks linked to a milestone with a deadline in the past. This filter is also added in the Burndown chart and Tasks analysis report. 
- When the user set a milestone on a task and no SOL is found in the project and/or parent task related then set the SOL of the milestone as a default SOL for the task.
- Display in green the milestone in project update view if all tasks linked to that milestone are in a closed stage.
- When a user changes a stage of a task to a closed stage, a wizard will be displayed if the task is linked to a milestone and all others tasks are in a closed stage too. This wizard will allow the user to reach the milestone since all tasks linked to that milestone are considered as finished.
- Remove `delivered_milestone` option in the `invoice_policy` selection field in `product.template` model when the Milestones feature is globally disabled. If some products are milestone one than they will be converted into a manual service ones to allow the user to manually set the quantity delivered for those products.
- Display the milestone field in task in Project Sharing feature.

task-2829542

### CODE REVIEW

File: addons/sale_project/models/product.py [hunk 1]
```diff
@@ -5,17 +5,21 @@
...
+    @api.model
+    def _selection_service_policy(self):
+        service_policies = [
+            # (service_policy, string)
+            ('ordered_prepaid', 'Prepaid/Fixed Price'),
+            ('delivered_milestones', 'Based on Milestones'),
+            ('delivered_manual', 'Based on Delivered Quantity (Manual)'),
+        ]
+        if not self.user_has_groups('project.group_project_milestone'):
+            service_policies = [policy for policy in service_policies if policy[0] != 'delivered_milestones']
```
Comment (MissingNoShiny) [inline addons/sale_project/models/product.py:20]: Wouldn't it be cleaner like this?
```suggestion
        service_policies = [
            # (service_policy, string)
            ('ordered_prepaid', 'Prepaid/Fixed Price'),
            ('delivered_manual', 'Based on Delivered Quantity (Manual)'),
        ]
        if self.user_has_groups('project.group_project_milestone'):
            service_policies.insert(1, ('delivered_milestones', 'Based on Milestones'))
```

File: addons/project/models/project.py [hunk 2]
```diff
@@ -1088,6 +1090,20 @@ def _read_group_personal_stage_type_ids(self, stages, domain, order):
...
+    allow_milestones = fields.Boolean(related='project_id.allow_milestones')
+    milestone_id = fields.Many2one(
+        'project.milestone',
+        'Milestone',
+        domain="[('project_id', '=', project_id)]",
+        compute='_compute_milestone_id',
+        readonly=False,
+        store=True,
+    )
+    has_late_and_not_reached_milestone = fields.Boolean(
```
Comment (MissingNoShiny) [inline addons/project/models/project.py:1102]: Maybe a bit nitpicky, but I would replace the occurrences of `/not_reached(?=_milestones?)/` with the proper word `unreached`, to make it clear that this isn't the negation of another value.

File: addons/project/report/project_task_burndown_chart_report_views.xml [hunk 3]
```diff
@@ -17,10 +18,12 @@
                 <filter name="filter_date_assign" date="date_assign"/>
                 <filter string="Last Month" invisible="1" name="last_month" domain="[('date','&gt;=', (context_today() - datetime.timedelta(days=30)).strftime('%Y-%m-%d'))]"/>
                 <filter string="Open tasks" name="open_tasks" domain="[('is_closed', '=', False)]"/>
+                <filter string="Late Milestones" name="late_milestone" domain="[('task_id.has_late_and_not_reached_milestone', '=', True)]" groups="project.group_project_milestone"/>
                 <group expand="0" string="Group By">
                     <filter string="Date" name="date" context="{'group_by': 'date'}" />
                     <filter string="Stage" name="stage" context="{'group_by': 'stage_id'}" />
                     <filter string="Project" name="project" context="{'group_by': 'project_id'}" />
+                    <filter string="Milestone" name="milestone" context="{'group_by': 'milestone_id'}" />
```
Comment (MissingNoShiny) [inline addons/project/report/project_task_burndown_chart_report_views.xml:26]:
```suggestion
                    <filter string="Milestone" name="milestone" context="{'group_by': 'milestone_id'}" groups="project.group_project_milestone"/>
```
I think we need the group here too

File: addons/project/models/project.py [hunk 4]
```diff
@@ -2217,6 +2265,29 @@ def _get_all_subtasks(self, depth=0):
             return children
         return children + children._get_all_subtasks(depth - 1)
 
+    def get_milestone_to_mark_as_done_message(self):
+        """ Return an action if the milestone can be marked as done since all tasks is done """
+        self.ensure_one()
+        if self.milestone_id and self.milestone_id.can_be_marked_as_done:
+            return _('All tasks linked to the %s milestone are now closed, the milestone can be marked as reached.', self.milestone_id.display_name)
+        return False
```
[the remark this answers is not in the document]
Comment (xavierbol) [reply addons/project/models/project.py:2273]: Indeed, I had to remove it in my squash commit. 😓

File: addons/project/models/project.py (hunk 2 above)
Comment (xavierbol) [reply addons/project/models/project.py:1102]: Indeed, I agree, thanks. 🙂

File: addons/project/views/project_views.xml [hunk 5]
```diff
@@ -387,7 +390,7 @@
                             name="%(act_project_project_2_project_task_all)d" icon="fa-tasks">
                             <field string="Tasks In Progress" name="task_count" widget="statinfo" options="{'label_field': 'label_tasks'}"/>
                         </button>
-                        <button class="oe_stat_button" name="action_get_list_view" type="object" icon="fa-check-square-o">
+                        <button class="oe_stat_button" name="action_get_list_view" type="object" icon="fa-check-square-o" attrs="{'invisible': [('allow_milestones', '=', False)]}" groups="project.group_project_milestone">
```
Comment (MissingNoShiny) [inline addons/project/views/project_views.xml:393]:
```suggestion
                        <button class="oe_stat_button" name="action_get_list_view" type="object" icon="fa-check-square-o" attrs="{'invisible': [('allow_milestones', '=', False)]}">
```
I think the group is unnecessary here as it is already taken into account when computing `allow_milestones`

File: addons/project/views/project_views.xml [hunk 6]
```diff
@@ -1394,7 +1438,9 @@
                           event_open_popup="true" quick_add="false" show_unusual_days="True"
                           js_class="project_calendar"
                           scales="month,year">
+                    <field name="allow_milestones" invisible="1" />
                     <field name="project_id" widget="project_private_task"/>
+                    <field name="milestone_id" attrs="{'invisible': [('allow_milestones', '=', False)]}" groups="project.group_project_milestone"/>
```
Comment (MissingNoShiny) [inline addons/project/views/project_views.xml:1443]:
```suggestion
                    <field name="milestone_id" attrs="{'invisible': [('allow_milestones', '=', False)]}"/>
```
I think the group is unnecessary here as it is already taken into account when computing `allow_milestones`

File: addons/project/views/project_views.xml [hunk 7]
```diff
@@ -1056,6 +1071,12 @@
                             <field name="project_id" attrs="{'invisible': [('parent_id', '!=', False)]}" domain="[('active', '=', True), ('company_id', '=', company_id)]"
                                 placeholder="Private" class="o_project_task_project_field oe_edit_only"/>
                             <field name="display_project_id" string="Project" attrs="{'invisible': [('parent_id', '=', False)]}" domain="[('active', '=', True), ('company_id', '=', company_id)]"/>
+                            <field name="milestone_id"
+                                placeholder="e.g. Product Launch"
+                                context="{'default_project_id': project_id if not parent_id or not display_project_id else display_project_id}"
+                                attrs="{'invisible': ['|', ('project_id', '=', False), ('allow_milestones', '=', False)]}"
+                                groups="project.group_project_milestone"
+                            />
```
Comment (MissingNoShiny) [inline addons/project/views/project_views.xml:1079]:
```suggestion
                            <field name="milestone_id"
                                placeholder="e.g. Product Launch"
                                context="{'default_project_id': project_id if not parent_id or not display_project_id else display_project_id}"
                                attrs="{'invisible': ['|', ('project_id', '=', False), ('allow_milestones', '=', False)]}"
                            />
```
I think the group is unnecessary here as it is encompassed by `allow_milestones`

File: addons/project/wizard/project_milestone_reach_wizard.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+    def action_mark_milestone_as_reach(self):
+        milestones = self.line_ids.filtered('mark_as_reached').milestone_id if self.line_count > 1 else self.milestone_id
+        if not milestones:
+            return {'type': 'ir.actions.act_window_close'}
+        milestones.is_reached = True
+        if len(milestones) == 1:
+            message = _('The %s milestone has successfully been marked as reached.', milestones.display_name)
+        else:
+            message = _('The milestones selected has successfully been marked as reached.')
```
Comment (MissingNoShiny) [inline addons/project/wizard/project_milestone_reach_wizard.py:28]:
```suggestion
            message = _('The selected milestones have successfully been marked as reached.')
```

File: addons/project/wizard/project_milestone_reach_wizard.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import _, api, fields, models
+
+
+class ProjectMilestoneReachWizard(models.TransientModel):
+    _name = 'project.milestone.reach.wizard'
+    _description = 'Mark a or more milestones as reached'
```
Comment (MissingNoShiny) [inline addons/project/wizard/project_milestone_reach_wizard.py:9]:
```suggestion
    _description = 'Mark one or more milestones as reached'
```

**Review by MissingNoShiny — CHANGES_REQUESTED** [review]
Neat features, thanks for your work. I have made a few suggestions, let me know what you think :slightly_smiling_face:

File: addons/project/wizard/project_milestone_reach_wizard.py (hunk 9 above)
Comment (xavierbol) [reply addons/project/wizard/project_milestone_reach_wizard.py:9]: 🙈

**Review by IT-Ideas — COMMENTED** [review]
Thanks for the work done!
It looks great 🤩

File: addons/project/data/project_demo.xml [hunk 10]
```diff
@@ -204,6 +204,27 @@
...
+            <field name="deadline" eval="(DateTime.now() + relativedelta(years=1)).strftime('%Y-%m-15')"/>
+            <field name="name">Second Phase</field>
+            <field name="project_id" ref="project.project_project_1"/>
+        </record>
+        <record id="project_milestone_3" model="project.milestone">
+            <field name="is_reached" eval="False"/>
+            <field name="deadline" eval="(DateTime.now() + relativedelta(years=2)).strftime('%Y-%m-%d')"/>
+            <field name="name">Final Phase</field>
+            <field name="project_id" ref="project.project_project_1"/>
+        </record>
```
[the remark this answers is not in the document]
Comment (xavierbol) [reply addons/project/data/project_demo.xml:231]: Good idea! 🙂

File: addons/project/report/project_task_burndown_chart_report_views.xml [hunk 11]
```diff
@@ -17,10 +18,12 @@
                 <filter name="filter_date_assign" date="date_assign"/>
                 <filter string="Last Month" invisible="1" name="last_month" domain="[('date','&gt;=', (context_today() - datetime.timedelta(days=30)).strftime('%Y-%m-%d'))]"/>
                 <filter string="Open tasks" name="open_tasks" domain="[('is_closed', '=', False)]"/>
+                <filter string="Late Milestones" name="late_milestone" domain="[('task_id.has_late_and_unreached_milestone', '=', True)]" groups="project.group_project_milestone"/>
```
[the remark this answers is not in the document]
Comment (xavierbol) [reply addons/project/report/project_task_burndown_chart_report_views.xml:21]: Indeed, thanks. 🙂

**Review by IT-Ideas — APPROVED** [review]
Thanks 🙂

**Comment by IT-Ideas** [comment]
You can squash your commits 🙂

**Review by IT-Ideas — APPROVED** [review]
Thanks 🙂 
@robodoo merge r+

**Pull request merged**